includes/class-magic-link.php

```diff
@@ -231,6 +265,56 @@ public static function clear_user_tokens( $user ) {
 		do_action( 'newspack_magic_link_user_tokens_cleared', $user );
 	}
 
+	/**
+	 * Get a random OTP code.
+	 *
+	 * @return string Random OTP code.
+	 */
+	private static function get_random_otp_code() {
+		$length = self::OTP_LENGTH;
+		$otp    = '';
+		for ( $i = 0; $i < $length; $i++ ) {
+			$otp .= strval( \wp_rand( 0, 9 ) );
+		}
+		return $otp;
+	}
+
+	/**
+	 * Generate an OTP to be used as part of a token.
+	 *
+	 * @param \WP_User $user User to generate OTP for.
+	 *
+	 * @return array|null {
+	 *   OTP or null if unable to generate one.
+	 *
+	 *   @type string $code     OTP code.
+	 *   @type string $hash     OTP Hash.
+	 *   @type int    $attempts Initial value for OTP authentication attempts.
+	 * }
+	 */
+	private static function generate_otp( $user ) {
+		if ( ! self::is_otp_enabled( $user ) ) {
+			return null;
+		}
+		$code = self::get_random_otp_code();
+		$hash = \wp_generate_password( 60, false, false );
+
+		/** This filter is documented in wp-includes/pluggable.php */
+		if ( \apply_filters( 'send_auth_cookies', true ) ) {
+			/** Subtract 1 minute to the OTP hash cookie expiration. */
```

Comment:
Why should one minute be subtracted?

Reply:
The OTP hash is attached to the token, so once the token expires, the OTP hash is also invalid.

The existence of an OTP hash cookie determines the "Sign In" modal UI to enter an OTP code. If the token is about to expire, it's best that the OTP hash is no longer available so the reader doesn't have a chance to attempt a request that is guaranteed to fail.